File: Android/app/src/main/java/com/example/liquorapplication/AuthenticationManager.kt
package com.example.liquorapplication

import android.content.Context
import android.util.Log
import com.couchbase.lite.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * Authentication Manager using Couchbase Lite for session persistence
 * Stores user session in local database to maintain login state across app restarts
 */
class AuthenticationManager(
    private val context: Context,
    private val databaseManager: DatabaseManager
) {
    
    companion object {
        private const val TAG = "AuthManager"
        private const val SESSION_DOC_ID = "user_session"
        private const val AUTH_COLLECTION = "auth"
    }
    
    private var database: Database? = null
    var currentUser: User? = null
        private set
    var isAuthenticated: Boolean = false
        private set
    var storeProfile: StoreProfile? = null
        private set
    
    // Valid credentials (matches iOS - only these 2 credentials)
    private val validCredentials = mapOf(
        // NYC user → NYC endpoint (supermarket-nyc)
        "nyc-store-01@supermarket.com" to UserCredentials(
            password = "P@ssword1",
            fullName = "NYC Store Manager",
            role = "Store Manager",
            endpoint = "supermarket-nyc"
        ),
        // AA user → AA endpoint (supermarket-aa)
        "aa-store-01@supermarket.com" to UserCredentials(
            password = "P@ssword1",
            fullName = "Ann Arbor Store Manager",
            role = "Store Manager",
            endpoint = "supermarket-aa"
        )
    )
    
    init {
        initializeDatabase()
        checkStoredLogin()
    }
    
    private fun initializeDatabase() {
        try {
            // Initialize Couchbase Lite if not already initialized
            CouchbaseLite.init(context)
            
            // Open or create auth database
            val dbConfig = DatabaseConfiguration()
            database = Database("AuthDB", dbConfig)
            
            Log.d(TAG, "✅ Authentication database initialized")
        } catch (e: Exception) {
            Log.e(TAG, "❌ Failed to initialize auth database", e)
        }
    }
    
    /**
     * Authenticate user with username and password
     */
    suspend fun login(username: String, password: String): LoginResult = withContext(Dispatchers.IO) {
        try {
            // Validate credentials
            val credentials = validCredentials[username.lowercase()]
            
            if (credentials == null || credentials.password != password) {
                return@withContext LoginResult.Error("Invalid username or password")
            }
            
            // Fetch store profile from database
            storeProfile = databaseManager.getStoreProfile()
            
            // Create user object with profile name
            val user = User(
                username = username.lowercase(),
                fullName = credentials.fullName,
                role = credentials.role,
                profileName = storeProfile?.name  // Use profile name from Capella
            )
            
            // Store session in Couchbase Lite
            storeSession(user)
            
            // Update state
            currentUser = user
            isAuthenticated = true
            
            Log.d(TAG, "✅ Login successful: ${user.fullName}")
            Log.d(TAG, "📋 Profile name: ${storeProfile?.name ?: "Not loaded"}")
            return@withContext LoginResult.Success(user)
            
        } catch (e: Exception) {
            Log.e(TAG, "❌ Login error", e)
            return@withContext LoginResult.Error("Login failed: ${e.message}")
        }
    }
    
    /**
     * Logout current user and clear session
     */
    fun logout() {
        try {
            clearSession()
            currentUser = null
            isAuthenticated = false
            Log.d(TAG, "🚪 User logged out")
        } catch (e: Exception) {
            Log.e(TAG, "❌ Logout error", e)
        }
    }
    
    /**
     * Check for stored login session on app start
     */
    private fun checkStoredLogin() {
        try {
            val collection = database?.defaultCollection ?: return
            val sessionDoc = collection.getDocument(SESSION_DOC_ID)
            
            if (sessionDoc != null) {
                val username = sessionDoc.getString("username")
                val fullName = sessionDoc.getString("fullName")
                val role = sessionDoc.getString("role")
                val profileName = sessionDoc.getString("profileName")
                val isAuth = sessionDoc.getBoolean("isAuthenticated")
                
                if (isAuth && username != null && fullName != null && role != null) {
                    // Validate that the stored credentials still match the current app configuration
                    val storedCredentials = validCredentials[username]
                    if (storedCredentials == null) {
                        // Credentials no longer valid, clear session
                        Log.d(TAG, "⚠️ Stored credentials are no longer valid, clearing session")
                        clearSession()
                        return
                    }
                    
                    // Validate that the stored session matches the current store configuration
                    val expectedUsername = when (AppConfig.currentStore) {
                        StoreLocation.AA -> "aa-store-01@supermarket.com"
                        StoreLocation.NYC -> "nyc-store-01@supermarket.com"
                    }
                    
                    if (username != expectedUsername) {
                        Log.d(TAG, "⚠️ Stored session ($username) doesn't match current store (${AppConfig.currentStore.displayName}), clearing session")
                        clearSession()
                        return
                    }
                    
                    // Restore session
                    currentUser = User(username, fullName, role, profileName)
                    isAuthenticated = true
                    Log.d(TAG, "🔄 Restored login session: $fullName")
                } else {
                    Log.d(TAG, "ℹ️ No valid stored session found")
                }
            } else {
                Log.d(TAG, "ℹ️ No stored session document")
            }
        } catch (e: Exception) {
            Log.e(TAG, "❌ Error checking stored login", e)
            // Clear invalid session on error
            clearSession()
        }
    }
    
    /**
     * Store user session in Couchbase Lite
     */
    private fun storeSession(user: User) {
        try {
            val collection = database?.defaultCollection ?: return
            
            val sessionDoc = MutableDocument(SESSION_DOC_ID)
            sessionDoc.setString("username", user.username)
            sessionDoc.setString("fullName", user.fullName)
            sessionDoc.setString("role", user.role)
            sessionDoc.setString("profileName", user.profileName)
            sessionDoc.setBoolean("isAuthenticated", true)
            sessionDoc.setLong("loginTime", System.currentTimeMillis())
            
            collection.save(sessionDoc)
            Log.d(TAG, "💾 Session stored in Couchbase Lite")
        } catch (e: Exception) {
            Log.e(TAG, "❌ Error storing session", e)
        }
    }
    
    /**
     * Clear user session from Couchbase Lite
     */
    private fun clearSession() {
        try {
            val collection = database?.defaultCollection ?: return
            val sessionDoc = collection.getDocument(SESSION_DOC_ID)
            
            if (sessionDoc != null) {
                collection.delete(sessionDoc)
                Log.d(TAG, "🗑️  Session cleared from Couchbase Lite")
            }
        } catch (e: Exception) {
            Log.e(TAG, "❌ Error clearing session", e)
        }
    }
    
    /**
     * Get all demo users for credentials screen
     */
    fun getAllUsers(): List<DemoUser> {
        return validCredentials.map { (username, creds) ->
            DemoUser(
                username = username,
                fullName = creds.fullName,
                role = creds.role,
                endpoint = creds.endpoint,
                password = creds.password
            )
        }.sortedBy { it.username }
    }
    
    /**
     * Close database connection
     */
    fun close() {
        try {
            database?.close()
            database = null
        } catch (e: Exception) {
            Log.e(TAG, "Error closing auth database", e)
        }
    }
}

/**
 * User data model
 */
data class User(
    val username: String,
    val fullName: String,
    val role: String,
    val profileName: String? = null  // From profile collection in Capella
)

/**
 * User credentials data model
 */
private data class UserCredentials(
    val password: String,
    val fullName: String,
    val role: String,
    val endpoint: String
)

/**
 * Demo user info for credentials screen
 */
data class DemoUser(
    val username: String,
    val fullName: String,
    val role: String,
    val endpoint: String,
    val password: String
)

/**
 * Login result sealed class
 */
sealed class LoginResult {
    data class Success(val user: User) : LoginResult()
    data class Error(val message: String) : LoginResult()
}

File: iOS/LiquorApp/AuthenticationManager.swift
import Foundation
import SwiftUI
import CouchbaseLiteSwift

// MARK: - User Model
struct User {
    let username: String
    let fullName: String
    let role: String
}

// MARK: - Authentication Manager with Couchbase Lite Session Persistence
class AuthenticationManager: ObservableObject {
    @Published var isAuthenticated = false
    @Published var currentUser: User?
    @Published var loginError: String?
    @Published var showingLoginError = false
    
    private var database: Database?
    private let sessionDocID = "user_session"
    
    // Capella App Services credentials
    // Only these 2 credentials are valid
    private let validCredentials: [String: (password: String, fullName: String, role: String, endpoint: String)] = [
        // NYC user → NYC endpoint (supermarket-nyc)
        "nyc-store-01@supermarket.com": (
            password: "P@ssword1",
            fullName: "NYC Store Manager",
            role: "Store Manager",
            endpoint: "supermarket-nyc"
        ),
        // AA user → AA endpoint (supermarket-aa)
        "aa-store-01@supermarket.com": (
            password: "P@ssword1",
            fullName: "Ann Arbor Store Manager",
            role: "Store Manager",
            endpoint: "supermarket-aa"
        )
    ]
    
    init() {
        // Initialize Couchbase Lite database for session storage
        initializeDatabase()
        // Check if user is already logged in (session persistence)
        checkStoredLogin()
    }
    
    // MARK: - Database Initialization
    
    private func initializeDatabase() {
        do {
            // Create or open authentication database
            self.database = try Database(name: "AuthDB")
            print("✅ Authentication database initialized")
        } catch {
            print("❌ Failed to initialize auth database: \(error)")
        }
    }
    
    // MARK: - Authentication Methods
    
    func login(username: String, password: String) {
        // Ensure we're on the main thread for UI updates
        DispatchQueue.main.async {
            // Clear any previous error
            self.loginError = nil
            
            // Validate credentials
            guard let userCredentials = self.validCredentials[username.lowercased()],
                  userCredentials.password == password else {
                self.loginError = "Invalid username or password"
                return
            }
            
            // Create user object
            let user = User(
                username: username.lowercased(),
                fullName: userCredentials.fullName,
                role: userCredentials.role
            )
            
            // Update authentication state
            withAnimation(.easeInOut(duration: 0.3)) {
                self.currentUser = user
                self.isAuthenticated = true
            }
            
            // Store login state for session persistence
            self.storeLoginState(user: user)
            
            print("✅ Login successful: \(user.fullName) (\(user.role))")
        }
    }
    
    func logout() {
        withAnimation(.easeInOut(duration: 0.3)) {
            self.isAuthenticated = false
            self.currentUser = nil
            self.loginError = nil
        }
        
        // Clear stored login state
        clearStoredLogin()
        
        print("🚪 User logged out")
    }
    
    // MARK: - Session Persistence with Couchbase Lite
    
    private func storeLoginState(user: User) {
        guard let database = database else {
            print("❌ Database not initialized")
            return
        }
        
        do {
            let sessionDoc = MutableDocument(id: sessionDocID)
            sessionDoc.setString(user.username, forKey: "username")
            sessionDoc.setString(user.fullName, forKey: "fullName")
            sessionDoc.setString(user.role, forKey: "role")
            sessionDoc.setBoolean(true, forKey: "isAuthenticated")
            sessionDoc.setDate(Date(), forKey: "loginTime")
            
            try database.saveDocument(sessionDoc)
            print("💾 Session stored in Couchbase Lite")
        } catch {
            print("❌ Error storing session: \(error)")
        }
    }
    
    private func checkStoredLogin() {
        guard let database = database,
              let sessionDoc = database.document(withID: sessionDocID) else {
            print("ℹ️ No stored session found")
            return
        }
        
        guard sessionDoc.boolean(forKey: "isAuthenticated"),
              let username = sessionDoc.string(forKey: "username"),
              let fullName = sessionDoc.string(forKey: "fullName"),
              let role = sessionDoc.string(forKey: "role") else {
            print("ℹ️ Invalid session data")
            return
        }
        
        // Validate that the stored credentials still match the current app configuration
        guard validCredentials[username] != nil else {
            print("⚠️ Stored credentials are no longer valid, clearing session")
            clearStoredLogin()
            return
        }
        
        // Validate that the stored session matches the current store configuration
        let expectedUsername: String
        switch AppConfig.currentStore {
        case .aa:
            expectedUsername = "aa-store-01@supermarket.com"
        case .nyc:
            expectedUsername = "nyc-store-01@supermarket.com"
        }
        
        guard username == expectedUsername else {
            print("⚠️ Stored session (\(username)) doesn't match current store (\(AppConfig.currentStore.displayName)), clearing session")
            clearStoredLogin()
            return
        }
        
        // Restore user session
        let user = User(username: username, fullName: fullName, role: role)
        DispatchQueue.main.async {
            self.currentUser = user
            self.isAuthenticated = true
        }
        
        print("🔄 Restored login session from Couchbase Lite: \(user.fullName)")
    }
    
    private func clearStoredLogin() {
        guard let database = database,
              let sessionDoc = database.document(withID: sessionDocID) else {
            return
        }
        
        do {
            try database.deleteDocument(sessionDoc)
            print("🗑️ Session cleared from Couchbase Lite")
        } catch {
            print("❌ Error clearing session: \(error)")
        }
    }
    
    // MARK: - Utility Methods
    
    func getAllUsers() -> [(username: String, fullName: String, role: String, endpoint: String, password: String)] {
        return validCredentials.map { (username, details) in
            (username: username, fullName: details.fullName, role: details.role, endpoint: details.endpoint, password: details.password)
        }.sorted { $0.username < $1.username }
    }
}
